### PR TITLE
Bug pattern for ensuring JMock assertions are evaluated

### PR DESCRIPTION
**Pull request by tom-smalls**
JMock uses a Mockery object to mock objects and to declare expectations on said mocks. If the field that holds the mockery object is not declared as a JUnit rule, or if the test does not use the JMock test runner, then this expectations are not evaluated and your tests are now passing for the wrong reasons.

I've included a check for @RunWith(JMock.class) although this is now deprecated behaviour in the latest version of JMock, and several test cases for dealing with both @Rule and @RunWith.

Thanks,

Tom

### CODE REVIEW

**Comment by tom-smalls** [comment]
I forgot to mark this check as experimental and to put it into the disabled list

File: annotation/src/main/java/com/google/errorprone/BugPattern.java [hunk 1]
```diff
@@ -16,12 +16,12 @@
 
 package com.google.errorprone;
 
-import static java.lang.annotation.RetentionPolicy.RUNTIME;
-
 import java.lang.annotation.Annotation;
 import java.lang.annotation.Retention;
 import java.util.Comparator;
 
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
```
Comment (eaftan) [inline annotation/src/main/java/com/google/errorprone/BugPattern.java]: Can you revert this import ordering change?  Google style is for static imports to go together at the top:
https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s3.3.3-import-ordering-and-spacing

File: core/src/main/java/com/google/errorprone/bugpatterns/JMockTestWithoutRunWithOrRuleAnnotation.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+public class JMockTestWithoutRunWithOrRuleAnnotation extends BugChecker implements BugChecker.VariableTreeMatcher {
+
+    private static final Matcher<VariableTree> JMOCK_MOCKERY_MATCHER = allOf(isSameType("org.jmock.Mockery"), isField());
+    private static final Matcher<VariableTree> FIELD_A_RULE_ANNOTATION_MATHER = hasAnnotation("org.junit.Rule");
+    private static final Matcher<Tree> CLASS_USING_JMOCK_RUNNER_MATCHER = enclosingClass(Matchers.<ClassTree>annotations(ChildMultiMatcher.MatchType.ALL, allOf(isType("org.junit.runner.RunWith"),
+            hasArgumentWithValue("value", classLiteral(isSameType("org.jmock.integration.junit4.JMock"))))));
+
+    @Override
+    public Description matchVariable(VariableTree tree, VisitorState state) {
+        if (JMOCK_MOCKERY_MATCHER.matches(tree, state) && !(FIELD_A_RULE_ANNOTATION_MATHER.matches(tree, state) || CLASS_USING_JMOCK_RUNNER_MATCHER.matches(tree, state))) {
```
Comment (eaftan) [inline core/src/main/java/com/google/errorprone/bugpatterns/JMockTestWithoutRunWithOrRuleAnnotation.java]: I might restructure this as:

```
allOf(
    JMOCK_MOCKERY_MATCHER,
    not(anyOf(FIELD_A_RULE_ANNOTATION_MATCHER, CLASS_USING_JMOCK_RUNNER_MATCHER)))
```

Extract that composite matcher out to a field, and then use that here.

File: core/src/main/java/com/google/errorprone/bugpatterns/JMockTestWithoutRunWithOrRuleAnnotation.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+@BugPattern(name = "JMockTestWithoutRunWithOrRuleAnnotation",
+        summary = "JMock tests must have @RunWith class annotation or the mockery field declared as a JUnit rule",
+        explanation = "If this is not done then all of your JMock tests will run and pass. However none of your assertions " +
+                "will actually be evaluated so your tests will be producing false positive results.",
+        category = Category.JMOCK, severity = ERROR, maturity = EXPERIMENTAL)
+public class JMockTestWithoutRunWithOrRuleAnnotation extends BugChecker implements BugChecker.VariableTreeMatcher {
+
+    private static final Matcher<VariableTree> JMOCK_MOCKERY_MATCHER = allOf(isSameType("org.jmock.Mockery"), isField());
+    private static final Matcher<VariableTree> FIELD_A_RULE_ANNOTATION_MATHER = hasAnnotation("org.junit.Rule");
```
Comment (eaftan) [inline core/src/main/java/com/google/errorprone/bugpatterns/JMockTestWithoutRunWithOrRuleAnnotation.java]: nit: MATHER should be MATCHER

File: core/src/main/java/com/google/errorprone/scanner/BuiltInCheckerSuppliers.java [hunk 4]
```diff
@@ -20,105 +20,8 @@
...
-import com.google.errorprone.bugpatterns.UnnecessaryTypeArgument;
-import com.google.errorprone.bugpatterns.WaitNotInLoop;
-import com.google.errorprone.bugpatterns.WildcardImport;
-import com.google.errorprone.bugpatterns.WrongParameterPackage;
-import com.google.errorprone.bugpatterns.threadsafety.GuardedByChecker;
-import com.google.errorprone.bugpatterns.threadsafety.GuardedByValidator;
-import com.google.errorprone.bugpatterns.threadsafety.LockMethodChecker;
-import com.google.errorprone.bugpatterns.threadsafety.SynchronizeOnNonFinalField;
-import com.google.errorprone.bugpatterns.threadsafety.UnlockMethodChecker;
+import com.google.errorprone.bugpatterns.*;
```
Comment (eaftan) [inline core/src/main/java/com/google/errorprone/scanner/BuiltInCheckerSuppliers.java]: Google style doesn't allow wildward imports, so please revert these lines.
https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s3.3.1-wildcard-imports

**Comment by tom-smalls** [comment]
I'll cancel the pull request and create a new one once I've sorted out all the styling issues

**Comment by eaftan** [comment]
Thanks for the contribution!  I haven't used JMock, so I'll take your word for it that the bug pattern is correct.  Seems like a good thing to check.

File: core/src/main/java/com/google/errorprone/scanner/BuiltInCheckerSuppliers.java (hunk 4 above)
Comment (tom-smalls) [inline core/src/main/java/com/google/errorprone/scanner/BuiltInCheckerSuppliers.java]: At the moment you have .idea set in .gitignore. Can we change this to be something a bit more? It means the intellij xml files that dictate things like code styling etc could be committed

**Comment by tom-smalls** [comment]
I have reopened the pull request with the requested style and refactoring changes.

I also changed the JMock matcher to see if the field is castable to Mockery as there are several implementations (most notable JUnitRuleMockery) that extend mockery that should be picked up too.

**Comment by eaftan** [comment]
Quick question: What if you are using JUnit3?  Could this checker flag legitimate test code that uses JUnit 3?

**Comment by tom-smalls** [comment]
Rule's don't exist in JUnit3. I believe this is only applicable JUnit4 and JMock 2.7+ so I'm not sure how much value this will actually have to the extended community

**Comment by eaftan** [comment]
Sorry again for the delay.  I cleaned this up a bit and commited in 5b393acb32810bc45e960a5f83f51194d7237a37.

Thanks for the contribution!

**Pull request closed without merging**